### DISCUSSION TITLE
How to get both type checking and JavaScript output?

### DISCUSSION THREAD

**Discussion by gwhitney** [DanielXMoore/Civet#678] [Q&A] [answered]
Although in retrospect it might have been naive, I was misled by the comment in the quickstart 

> To use TypeScript for type checking, create a tsconfig.json file. For example: ...

into thinking that as long as I had a (reasonable) tsconfig.json in my top-level directory, civet would check types whenever it compiled. So I just made my (entire) build script be `civet --js -c src/*civet -o dist/.js`. Quickly the absence of type errors disabused me of this notion.

But what is the best way to achieve both type checking and JavaScript output? I'm no real fan of TypeScript; I am here in fairly large part because CoffeeScript appears relatively "dead." But to the extent that I can and do make type annotations, I would like them to be enforced. Does Civet actually do no typechecking of its own? I.e., is the only way to get typechecking and end up with JavaScript to do a two-step build .civet -> .ts -> .js? (Is it the case that I will necessarily end up with the same .js if I do that as if I just execute `civet --js -c` ?) Or is there a way of configuring the civet command so that it will "automatically" arrange for the typechecking and then spit out the .js so I don't have to mess with any .ts files (well, except for .d.ts files, which I would like because possible clients of a putative library I would produce will want them to be there)?

Sorry if all these things should be clear from the existing documentation and I am just being too dense. Thanks!

**Comment by STRd6** [comment, marked as answer]
Don't worry about the questions, Civet is a fairly young project and there are still gaps in the tools and documentation. We're glad you're here!

Using the Civet LSP in VSCode (https://marketplace.visualstudio.com/items?itemName=DanielX.civet) adds development time type checking. It works pretty well but also has some known issues.

If you are not using VSCode then the two step `.civet` -> `.ts` -> `.js` is probably the only way currently. We plan to add to / upgrade our command line tools to do checking directly at some point (equivalent to svelte-check, vue-tsc, etc.) but I haven't gotten around to it yet. We also have plans for a generic non-vscode LSP at some point.

Thanks for starting this discussion. Civet is improving all the time and I try to focus on what components people need most when prioritizing work. As the community grows the tools and ecosystem will continue to as well.

**Reply by edemaine** [reply]
Just a few followup links:
* #61 tracks adding a CLI to do type checking (via Civet → ts → tsc), e.g. for CI
* Our recently added [unplugin](https://github.com/DanielXMoore/Civet/tree/main/integration/unplugin) probably enables type checking via some bundlers and the right mix of plugins. Perhaps @Mokshit06 knows a working combination? Does turning on `dts` mode suffice?

**Reply by gwhitney** [reply]
Thanks. I was a little unclear in #61 why there's such a clear consensus that `civetc` shouldn't both typecheck and compile all the way down to javascript, as that's exactly what I'd like to do. Perhaps optionally it could?

**Reply by edemaine** [reply]
Yeah, I had in mind that it would also generate JS via TypeScript, unless `noEmit` is true in tsconfig, just like tsc.

**Reply by edemaine** [reply]
Just to follow up on this, `civet --typecheck` now does typechecking at the CLI level. See https://civet.dev/getting-started#typechecking
